MushroomObserver/mushroom-observer
# Block Anonymous Users

Blocks anonymous users by redirecting most requests to `account/login`.
The purpose of the PR is to improve website response. 
The PR delivers [Pivotal #179794199](https://www.pivotaltracker.com/story/show/179794199), which has some more details.

### Exceptions
There are a bunch of exceptions to redirection:
- Whitelisted Robots. Requests which are identified as coming from robots are handled in `kick_out_robots`, which either allows whitelisted robots to access certain pages, or kicks out non-whitelisted robots.
- api, ajax, and graphql requests
- account, articles, publications, policy, and support controllers
- observer :ask_webmaster_question, :how_to_use, :intro, :turn_javascript_nil, :turn_javascript_off,  :turn_javascript_on

### Suggested Manual Test
- logout
- Try going to some of the allowed places, e.g., http://localhost:3000/observer/intro, http://localhost:3000/observer/how_to_use, http://localhost:3000/policy/privacy, http://localhost:3000/support/donate, http://localhost:3000/account/signup, http://localhost:3000/articles, http://localhost:3000/observer/ask_webmaster_question, http://localhost:3000/publications
Also try viewing some pages with different screen sizes browser development tools.
Expected result: The page should be displayed.
- Try any other page. Examples: http://localhost:3000/observer/list_rss_logs, http://localhost:3000/name/observation_index, http://localhost:3000/location/list_locations
Expected result: redirected to login
- Login from your last redirect to login
Expected result: Page you tried to access before logging in.
